torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -568,17 +572,28 @@ std::unordered_map<std::string, std::string> ProcessGroupAgent::getMetrics() {
   std::unordered_map<std::string, std::string> metrics;
   {
     std::unique_lock<std::mutex> lock(futureMutex_);
-    metrics["num_pending_requests"] = c10::to_string(futures_.size());
+    metrics[kNumPendingRequests] = c10::to_string(futures_.size());
+  }
+  metrics[kThreadPoolSize] = c10::to_string(threadPool_.size());
+  metrics[kNumIdleThreads] = c10::to_string(threadPool_.numAvailable());
+  for (const auto& metricInfo : metricsMap_) {
+    metrics[metricInfo.first] =
+        c10::to_string(metricInfo.second->currentAverage_);
   }
-  metrics["thread_pool_size"] = c10::to_string(threadPool_.size());
-  metrics["num_idle_threads"] = c10::to_string(threadPool_.numAvailable());
   return metrics;
 }
 
-std::unordered_map<std::string, std::string> ProcessGroupAgent::getDebugInfo() {
-  /* This would later include more info other than metrics for eg: may include
-     stack traces for the threads owned by the agent */
-  return getMetrics();
+void ProcessGroupAgent::addGilWaitTime(
+    const std::chrono::microseconds gilWaitTime) {
+  auto gilMetrics = metricsMap_.find(kGilAverageWaitTime);
```

Comment:
need to protect the map with a lock. This is a bit of work to do on each GIL acquisition... also, not sure how it dovetails with other work people have done in this area.

Since this is the only use case of the map, my preference here would be just to keep a standalone std::unique_ptr<> just for this, and avoid an expensive string-based table lookup for it each time. 

If you eventually want it to be more general/etc. you can just keep vector<unique_ptr<AverageMetricsTracker>>, where you have an integer-like stats enum (e.g. AVERAGE_WAIT_TIME=0, SOME_OTHER_STAT=1,...) to index into it, rather than string-based indexing. (It still is "general" since you have the key name in the struct, and can emit it in the metrics method.)

Reply:
Implemented this using a `vector<unique_ptr>` for now.